src/main/java/com/artipie/asto/cache/Cache.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 artipie.com
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.asto.cache;
+
+import com.artipie.asto.Content;
+import com.artipie.asto.Key;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Supplier;
+
+/**
+ * Generic reactive cache which returns cached content by key of exist or loads from remote and
+ * cache if doesn't exit.
+ *
+ * @since 0.24
+ */
+public interface Cache {
+
+    /**
+     * No cache, just load remote resource.
+     */
+    Cache NOP = (key, remote, ctl) -> remote.get();
+
+    /**
+     * Try to load content from cache or fallback to remote publisher if cached key doesn't exist.
+     * When loading remote item, the cache may save its content to the cache storage.
+     * @param key Cached item key
+     * @param remote Remote source
+     * @param control Cache control
+     * @return Content for key
+     */
+    CompletionStage<? extends Content> load(
+        Key key, Supplier<? extends CompletionStage<? extends Content>> remote,
+        CacheControl control
+    );
```

Comment:
@g4s8 design of this method seems a bit off: we pass a `Key` to the method and method always pass the same `Key` to `CacheControl` we have provided. Or there is scenario when we pass some other `Key` to the `CacheControl`? Cannot think of one. Maybe it would be better to specify `CacheControl` only once in constructor?

Reply:
@olegmoz I created this `CacheControl` to respect [cache-control HTTP directives](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control) sent by client: e.g. if we have some `Cache` instance for a `Slice`, and this slice receives two requests from the client: one request with `Cache-Control: max-age=<X>` header, and another with `Cache-Control: no-cache ` header, so in first case we need to check that cached item was cached `X` seconds at the latest, but for the seconds request we have to invalidate the cache. If we move it to constructor, then we need to instantiate cache object per request, but it seems not logically correct to create new cache for repository on request, WDYT?

Reply:
@g4s8 I see, but still do not see how `CacheControl` is a function of `Key`? Don't we always know value before invoking `load` method? Don't we always invoke `CacheControl` with the same `Key` we passed as first argument?

Reply:
@olegmoz the idea of this `CacheControl` class was to delegate responsibility of validating cached items to separate objects and to create unlimited amount of implementations depends on client app needs. Before I thought about creating `enum` with possible cache-control directives, but with this approach the implementation will be located in all  `Cache` implementations.